File: agent/library/src/main/java/org/matilda/messages/BinaryMessageSender.java
package org.matilda.messages;

import javax.inject.Inject;
import java.io.*;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;

public class BinaryMessageSender implements MessageSender {
    private final DataOutputStream mOutputStream;

    @Inject
    MessageSerializer mSerializer;

    @Inject
    public BinaryMessageSender(OutputStream outputStream) {
        mOutputStream = new DataOutputStream(outputStream);
    }

    @Override
    public void send(Message message) throws IOException {
        byte[] serializedMessage = mSerializer.serialize(message);
        sendLength(serializedMessage.length);
        sendData(serializedMessage);
    }

    private void sendLength(int length) throws IOException {
        byte[] array = ByteBuffer.allocate(Integer.BYTES)
                .order(ByteOrder.LITTLE_ENDIAN)
                .putInt(length)
                .array();
        mOutputStream.write(array);
    }

    private void sendData(byte[] bytes) throws IOException {
        mOutputStream.write(bytes);
    }
}

File: agent/library/src/main/java/org/matilda/messages/BinaryMessageSender.kt
package org.matilda.messages

import kotlinx.io.Sink
import kotlinx.io.writeIntLe
import me.tatarka.inject.annotations.Inject

@Inject
class BinaryMessageSender(private val mSink: Sink, private val mSerializer: MessageSerializer) : MessageSender {
    override fun send(message: Message) {
        val serializedMessage = mSerializer.serialize(message)
        sendLength(serializedMessage.size)
        sendData(serializedMessage)
        mSink.flush()
    }

    private fun sendLength(length: Int) {
        mSink.writeIntLe(length)
    }

    private fun sendData(bytes: ByteArray) {
        mSink.write(bytes)
    }
}
